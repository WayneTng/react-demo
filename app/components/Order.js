import React from 'react'
import {formatPrice} from '../lib/helpers'
import CSSTransitionGroup from 'react-addons-css-transition-group'
import Timer from './Timer'
import { observer } from 'mobx-react'
import orderStore from '../stores/Order'
import fishStore from '../stores/Fish'

@observer class Order extends React.Component {
  componentWillMount() {
    const { storeId } = this.props
    this.store = orderStore.getStore(storeId)
    this.fishStore = fishStore.getStore(storeId)
  }

  renderOrder(key) {
    const fish = this.fishStore.fishes[key]
    const count = this.store.order[key]
    const removeButton = <button onClick={() => this.store.removeFromOrder(key)}>&times;</button>

    if (!fish || fish.status === 'unavailable') {
      return <li key={key}>Sorry, Fish is no longer available! {removeButton}</li>
    }

    return (
      <li key={key}>
        <span>
          <CSSTransitionGroup
            component="span"
            className="count"
            transitionName="count"
            transitionEnterTimeout={250}
            transitionLeaveTimeout={250}
          >
            <span key={count}>{count}</span>
          </CSSTransitionGroup>

          lbs {fish.name} {removeButton}
        </span>
        <span className='price'> {formatPrice(count * fish.price)} </span>
      </li>
    )
  }

  render() {
    const orderIds = Object.keys(this.store.order)
    const total = orderIds.reduce((prevTotal, key) => {
      const fish = this.fishStore.fishes[key]
      const count = this.store.order[key]
      const isAvailable = fish && fish.status === 'available'
      if (isAvailable) {
        return prevTotal + (count * fish.price || 0)
      }
      return prevTotal
    }, 0)

    return(
      <div className='order-wrap'>
        <h2> Your Order </h2>
        <ul className='order'>
          {orderIds.map((key) => this.renderOrder(key))}
          <li className='total'>
            <strong> Total: </strong>
            {formatPrice(total)}
          </li>
        </ul>
        <Timer />
      </div>
    )
  }
}

Order.propTypes = {
  storeId: React.PropTypes.string.isRequired
}

export default Order
