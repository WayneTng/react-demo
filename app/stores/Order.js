import { observable } from 'mobx'

class OrderStore {
  @observable order = {}

  constructor(storeId) {
    this.storeId = storeId
    this.localStorageKey = `order-${this.storeId}`

    const localStorageRef = localStorage.getItem(this.localStorageKey)
    if (localStorageRef) {
      this.order = JSON.parse(localStorageRef)
    }
  }

  saveToLocalStorage() {
    localStorage.setItem(this.localStorageKey, JSON.stringify(this.order))
  }

  addToOrder(key) {
    this.order = {
      ...this.order,
      [key]: this.order[key] + 1 || 1
    }

    this.saveToLocalStorage()
  }

  removeFromOrder(key) {
    delete this.order[key]
    this.order = {...this.order}
    this.saveToLocalStorage()
  }
}

class OrderStoreFactory {
  stores = {}

  getStore(storeId) {
    if (!this.stores[storeId]) {
      this.stores[storeId] = new OrderStore(storeId)
    }

    return this.stores[storeId]
  }
}

const orderStoreFactory = new OrderStoreFactory()
export default orderStoreFactory
