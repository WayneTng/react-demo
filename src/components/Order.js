import React from 'react'

class Order extends React.Component {
  render() {
    return(
      <div>
        this is order
      </div>
    )
  }
}

export default Order
