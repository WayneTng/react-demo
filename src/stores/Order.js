import { observable } from 'mobx'

class OrderStore {
  @observable order = {}

  addToOrder(key) {
    this.order = {
      ...this.order,
      [key]: this.order[key] + 1 || 1
    }
  }

  removeFromOrder(key) {
    delete this.order[key]
    this.order = {...this.order}
  }
}

const singleton = new OrderStore()
export default singleton
