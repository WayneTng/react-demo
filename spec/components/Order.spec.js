import React from 'react'
import renderer from 'react-test-renderer'
import Order from 'components/Order'
import OrderStoreFactory from 'stores/Order'

jest.mock('react-dom')
jest.mock('react-addons-css-transition-group')

window.localStorage = {
  items: {},
  setItem: function(key, value) { this.items[key] = value },
  getItem: function(key) { return this.items[key] }
}

describe('<Order />', () => {
  it('renders the order list', () => {
    const store = OrderStoreFactory.getStore('testStore')
    store.addToOrder('fish1')
    store.addToOrder('fish2')

    const order = renderer.create(
      <Order storeId='testStore' />
    )

    const tree = order.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
