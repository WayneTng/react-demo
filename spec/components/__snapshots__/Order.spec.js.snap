exports[`<Order /> renders the order list 1`] = `
<div
  className="order-wrap">
  <h2>
     Your Order 
  </h2>
  <ul
    className="order">
    <li>
      Sorry, Fish is no longer available! 
      <button
        onClick={[Function onClick]}>
        ×
      </button>
    </li>
    <li>
      Sorry, Fish is no longer available! 
      <button
        onClick={[Function onClick]}>
        ×
      </button>
    </li>
    <li
      className="total">
      <strong>
         Total: 
      </strong>
      $0.00
    </li>
  </ul>
  <div>
    Seconds Elapsed: 
    0
  </div>
</div>
`;
